### PR TITLE
PR: Remove autosaves on successful file close to partially mitigate spurious creation bugs

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

Partially mitigates the impact of #9265 .

It was much more complex than initially anticipated, but with this PR Spyder now removes autosave files after successfully closing them (either choosing not to save, or with no changes present) if an autosave exists, either when closing individual files or when closing the editor entirely. In order to do the latter, since the action is not already synchronized between panes like closing files and the autosaves are not in the ``name_mapping`` dict of the first split pane, I just iterate through all the EditorStacks, ignoring errors (which will occur if there are more than two, or if another Spyder instance removes them first), to make sure all autosaves are removed.

I also added both unit (autosave) and functional (editor) tests of the ``remove_all_autosave_files()`` function I added, as well as improved the existing tests of the ``remove_autosave_file()`` function, and refactored the latter to take a filename directly rather than an finfo object to greatly simplify the former and because only the filename was ever actually used.

Note: This should *not* close the linked issue, because it doesn't resolve the root of the problem and there are other, less serious side effects this mitigation is inherently not able to address directly.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: CAM_Gerlach

### CODE REVIEW

File: spyder/plugins/editor/utils/autosave.py [hunk 1]
```diff
@@ -151,28 +151,42 @@ def create_unique_autosave_filename(self, filename, autosave_dir):
...
         del self.name_mapping[filename]
         self.stack.sig_option_changed.emit(
-                'autosave_mapping', self.name_mapping)
-        logger.debug('Removing autosave file %s', autosave_filename)
+            'autosave_mapping', self.name_mapping)
+        logger.debug('Removed autosave file %s', autosave_filename)
+
+    def remove_all_autosave_files(self, errors='raise'):
+        """
+        Remove all autosave files stored in this componenet's mapping.
```
Comment (ccordoba12) [inline spyder/plugins/editor/utils/autosave.py]: componenet's -> component's

File: spyder/plugins/editor/plugin.py [hunk 2]
```diff
@@ -377,6 +377,14 @@ def closing_plugin(self, cancelable=False):
             else:
                 for win in self.editorwindows[:]:
                     win.close()
+                # Remove autosave file on successful close to avoid #9265
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py]: ... to avoid *issue* 9265 ...

Comment (ccordoba12) [reply spyder/plugins/editor/plugin.py]: This is our convention to reference issues.

File: spyder/plugins/editor/widgets/editor.py [hunk 3]
```diff
@@ -1505,6 +1505,10 @@ def close_file(self, index=None, force=False):
 
             self.add_last_closed_file(finfo.filename)
 
+            # Remove autosave file on successful close to work around #9265
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py]: ... around *issue* 9265 ...

File: spyder/plugins/editor/plugin.py (hunk 2 above)
Comment (CAM-Gerlach) [reply spyder/plugins/editor/plugin.py]: Done

File: spyder/plugins/editor/widgets/editor.py (hunk 3 above)
Comment (CAM-Gerlach) [reply spyder/plugins/editor/widgets/editor.py]: Done

File: spyder/plugins/editor/utils/autosave.py (hunk 1 above)
Comment (CAM-Gerlach) [reply spyder/plugins/editor/utils/autosave.py]: I can't spell, lol...new feature request: spellchecker for docstrings.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Updated with the requested changes, thanks.

@jitseniesen I'd feel much better about my hack job if you could give this a review before it goes in, since autosave is your "baby"...

File: spyder/plugins/editor/widgets/editor.py (hunk 3 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/editor.py]: @cam-gerlach could you add this comment on hiw we reference issues somewhere on the wiki or even on the PR template ? 

Thanks! Great work

Comment (CAM-Gerlach) [reply spyder/plugins/editor/widgets/editor.py]: @goanpeca I reorganized the [Code style](https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style) page and added this along with a number of other standards and conventions

**Comment by jitseniesen** [comment]
This is fine for beta but I want to give it more thought for Spyder 4. I am afraid that this PR removes autosave files in some circumstances where we don't want to remove them. Specifically, what if `close_file()` is called with `force=True`, and what if a file is saved unsuccessfully?

**Comment by CAM-Gerlach** [comment]
> Specifically, what if `close_file()` is called with `force=True`, and what if a file is saved unsuccessfully?

@jitseniesen The only place this is currently done in the codebase that I could find is ``close_file_in_all_editorstacks``, which is called to close the file in split panes other than the one the user clicked after saving/closing the file (without force) and either successively saved it, chose not to save or the file was unmodified. If the user choses to cancel, or saving doesn't complete, this function shouldn't ever fire at all, whereas if they click no or the save completes (or the file is unmodified) the autosave should be removed as per the existing logic you implemented.

If we simply didn't remove the autosave if ``force=True``,the problem is the autosave doesn't exist in the name mapping of the first editorstack (only the other ones) so it presumably wouldn't get removed at all if the user clicked "close" in the left/top pane, whereas it seems it shouldn't be possible to delete it if saving were to not complete.

However, if further assurance is desired, I can just perform a pre-check in ``close_file`` before ``force``/``save_if_changed`` that the file is unmodified and only remove the autosave if that's the case, which should be mostly true if this bug is triggered (the only edge case I can think of right now would be modifying an unmodified file within the 60 second window, then closing it from the left pane). Of course, once #9265 is fixed, that won't be an issue anymore.

**Comment by jitseniesen** [comment]
> I can just perform a pre-check in `close_file` before `force`/`save_if_changed` that the file is unmodified

I don't think that is necessary.

**Pull request merged**